test/subscription_map.hpp

```diff
@@ -439,31 +439,39 @@ class multiple_subscription_map
     // Insert a key => value at the specified subscription path
     // returns the handle and true if key was inserted, false if key was updated
     template <typename V>
-    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key const &key, V&& value) {
+    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key&& key, V&& value) {
         auto path = this->find_subscription(subscription);
         if (path.empty()) {
             auto new_subscription_path = this->create_subscription(subscription);
-            new_subscription_path.back()->second.value[key] = std::forward<V>(value);
+            new_subscription_path.back()->second.value.insert(std::make_pair(std::forward<Key>(key), std::forward<V>(value)));
             ++this->map_size;
             return std::make_pair(this->path_to_handle(new_subscription_path), true);
         }
         else {
-            auto new_pair = std::make_pair(key, std::forward<V>(value));
-            auto insert_result = path.back()->second.value.insert(new_pair);
-            if(insert_result.second) {
+            // I have to do a find first, because insert_or_assign is available in C++17 only
+            // https://www.fluentcpp.com/2018/12/11/overview-of-std-map-insertion-emplacement-methods-in-cpp17/
+            auto& subscription_set = path.back()->second.value;
+            auto existing_iter = subscription_set.find(key);
+            if(existing_iter == subscription_set.end()) {
+                subscription_set.insert(std::make_pair(std::forward<Key>(key), std::forward<V>(value)));
                 this->increase_subscriptions(path);
                 ++this->map_size;
             } else {
-                insert_result.first->second = new_pair.second;
+                existing_iter->second = std::forward<V>(value);
             }
-            return std::make_pair(this->path_to_handle(path), insert_result.second);
+            return std::make_pair(this->path_to_handle(path), existing_iter == subscription_set.end());
         }
     }
 
+    template <typename V>
+    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key const &key, V&& value) {
```

Comment:
Why bother with this? It just adds an unnecessary extra copy of the Key class. The && version of the function should be sufficient on it's own.

Reply:
It is convenient, otherwise you would have to make a temporary when everytime you call insert_or_update with a const key.

Reply:
I'm really confused why you say this is the case. I don't see how it could be.

Can you elaborate?

Reply:
This won't compile, without the Key const & version:

````C++

    multiple_subscription_map<std::string, int> map;

    std::vector<std::string> values = {
        "example/test/A", "example/+/A", "example/#", "#"
    };

    for (auto const& i : values) {
        map.insert_or_update(i, i, 0);
    }

Reply:
Key&& should work. See the comment below.

Reply:
> Key&& should work. See the comment below.
> https://github.com/redboltz/mqtt_cpp/pull/713/files#r512493134

See https://github.com/redboltz/mqtt_cpp/pull/713/files#r512499437